### DISCUSSION TITLE
Does FastAPI Support Wildcard Media Types in Multipart Form Data?

### DISCUSSION THREAD

**Discussion by rickstaa** [fastapi/fastapi#12122] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, File, Form, UploadFile
from PIL import Image

app = FastAPI()

@app.post("/upload-image/")
async def upload_image(
    prompt: str = Form(...),
    image: UploadFile = File(..., description="Uploaded image to process. Accepts any image format.", media_type="image/*")
):
    # Open the uploaded image
    image = Image.open(image.file).convert("RGB")

    # Here, you would normally process the image. For now, we'll just return its details.
    return {
        "prompt": prompt,
        "filename": image.filename,
        "format": image.format,
        "size": image.size,
    }
```

### Description

Hi everyone,

I'm currently working on specifying multiple image file types as acceptable media types using the [File](https://fastapi.tiangolo.com/tutorial/request-files/) class in FastAPI. My objective is to ensure these media types are clearly documented in the Swagger UI, and subsequently, in my [Mintlify documentation](https://mintlify.com/) and [Speakeasy SDKs](https://www.speakeasy.com/openapi/paths/operations/content) for better API clarity. I’ve noticed that media type wildcards are supported by [Speakeasy](https://www.speakeasy.com/openapi/paths/operations/content).

I attempted to use a wildcard media type (e.g., `media_type="image/*"`), but it doesn't seem to be reflected in the generated OpenAPI Spec, and thus doesn't appear in the Swagger documentation. I'm generating the OpenAPI spec YAML using the [get_openapi](https://fastapi.tiangolo.com/how-to/extending-openapi/) utility. While the [OpenAPI documentation](https://swagger.io/docs/specification/describing-request-body) indicates that wildcard media types should be supported, it appears FastAPI may not support this feature yet.

Before I consider alternative approaches, I wanted to double-check if this functionality is already supported in FastAPI or if there's a known workaround that I might have missed.

Thank you in advance for your assistance!

### Operating System

Linux

### Operating System Details

Ubuntu 24.04

### FastAPI Version

0.111.0

### Pydantic Version

2.7.2

### Python Version

Python 3.11.9

**Comment by n8sty** [comment]
Yes it does. You need to leverage the additional response functionality described here: https://fastapi.tiangolo.com/advanced/additional-responses/#additional-media-types-for-the-main-response.

Adapting your example:

```
@app.post(
    "/upload-image/",
    responses={"200": {"content": {"image/*": {}, "description": "This works",}}}
)
async def upload_image(
    prompt: str = Form(...),
    image: UploadFile = File(..., description="Uploaded image to process. Accepts any image format.", media_type="image/*")
):
    # Open the uploaded image
    image = Image.open(image.file).convert("RGB")

    # Here, you would normally process the image. For now, we'll just return its details.
    return {
        "prompt": prompt,
        "filename": image.filename,
        "format": image.format,
        "size": image.size,
    }
```

**Reply by rickstaa** [reply]
@n8sty Thanks for your response! It's great to know that the output response is supported 👍🏻. However, I'm looking for a way to specify a wildcard `media_type` for a file-based input parameter. Here's what I have in mind:

```bash
image: Annotated[
    UploadFile,
    File(
        description="Uploaded image to modify with the pipeline.",
        media_type="image/*",
    ),
],
```

Do you know if there's a way to achieve this 🤔?

**Comment by yakubka** [comment]
FastAPI does not natively support wildcard media types in multipart form data declarations. You cannot declare `Accept: multipart/*` as a consumed type, and the generated OpenAPI spec will always show a specific media type for file upload endpoints.

If you need to accept files of any type without restricting by MIME type, declare the upload field as `bytes` or `UploadFile` without a `media_type` constraint, and validate the content type manually in the handler.

```python
from fastapi import UploadFile, HTTPException

@app.post("/upload")
async def upload(file: UploadFile):
    allowed = {"image/jpeg", "image/png", "application/pdf"}
    if file.content_type not in allowed:
        raise HTTPException(status_code=415, detail=f"Unsupported media type: {file.content_type}")
    contents = await file.read()
    return {"filename": file.filename, "size": len(contents)}
```

The `content_type` is read from the `Content-Type` part header supplied by the client. Note that clients can send any string in that header, so treat it as a hint for display purposes and validate the actual file bytes with a library like `python-magic` if security matters.
